Add bazel 0.25 release announcement (#162)
bazelbuild/bazel-blog: _posts/2019-06-1-bazel-0.25.md

---
layout: posts
title: "Bazel 0.25"
authors:
  - dkelmer
---

Bazel 0.25.0 has arrived with an exciting mix of new features, important bug fixes, and updated documentation.

This version of Bazel has been regression-tested against [over 50 open-source projects](https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/939). Want to get your open-source project on Bazel's CI to ensure that we catch any regressions or have suggestions what we should test against? Reach out to [bazel-dev@googlegroups.com](mailto:bazel-discuss@googlegroups.com) and we can set you up!

## Incompatible changes

The following incompatible flags are now enabled by default, and will be removed in a near-term release. Before updating to Bazel 0.25, you may first check if your codebase is compatible either by running [`bazelisk --migrate`](https://github.com/philwo/bazelisk) or by building your code with Bazel

> 

Reviewer: remove empty line

- [`--incompatible_remove_cpu_and_compiler_attributes_from_cc_toolchain`](https://github.com/bazelbuild/bazel/issues/7075) 

- [`--incompatible_use_toolchain_providers_in_java_common`](https://github.com/bazelbuild/bazel/issues/7186)

- [`--incompatible_remote_symlinks`](https://github.com/bazelbuild/bazel/issues/7917)

 

The following incompatible change flags have been added in this release and migration for them should begin. Check out the linked bugs to see when it is expected that the flags be turned on by default. 

- [`--incompatible_windows_native_test_wrapper`](https://github.com/bazelbuild/bazel/issues/6622)

- [`--incompatible_require_ctx_in_configure_features`](https://github.com/bazelbuild/bazel/issues/7793)

- [`--incompatible_do_not_split_linking_cmdline`](https://github.com/bazelbuild/bazel/issues/7687)

- [`--incompatible_string_join_requires_strings`](https://github.com/bazelbuild/bazel/issues/7802)

- [`--incompatible_use_python_toolchains`](https://github.com/bazelbuild/bazel/issues/7899) and see [#7375](https://github.com/bazelbuild/bazel/issues/7375)

## Android

>  - Android NDK C++ toolchain is now configured in Starlark

Reviewer 2: ```suggestion
 - Android NDK C++ toolchain is now configured in Starlark.
```

>  - Fix an issue where the Android resource processor did not surface errors from aapt2 compile and link actions.

Reviewer 2: ```suggestion
 - Fixed an issue where the Android resource processor did not surface errors from aapt2 compile and link actions.
```

## Other Changes

 - Bazel is now ~20MiB smaller, from unbundling the Android rules' runtime dependencies.

>  - genrules now support a $(RULEDIR) variable that resolves to the directory where the outputs of the rule are put.

Reviewer 2: ```suggestion
 - genrules now support a `$(RULEDIR)` variable that resolves to the directory where the outputs of the rule are stored.
```

 
 - genquery somepath output is now deterministic.
 

>  - Makes TreeArtifact deterministic.

Reviewer 2: ```suggestion
 - Tree artifacts are now deterministic.
```

It'll be nice to know what kind of non-determinism they had before.

>  - Don't drop the analysis cache when the same --define flag is set multiple times and the last value is the same (e.g. if the current invocation was run with "--define foo=bar" and the previous one was run with "--define foo=baz --define foo=bar").

Reviewer 2: ```suggestion
 - The analysis cache is no longer dropped when the same --define flag is set multiple times and the last value is the same (e.g. if the current invocation was run with "--define foo=bar" and the previous one was run with "--define foo=baz --define foo=bar").
```

This version of Bazel has been regression-tested against [over 50 open-source projects](https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/939). Want to get your open-source project on Bazel's CI to ensure that we catch any regressions or have suggestions what we should test against? Reach out to [bazel-dev@googlegroups.com](mailto:bazel-discuss@googlegroups.com) and we can set you up!

## Incompatible changes

The following incompatible flags are now enabled by default, and will be removed in a near-term release. Before updating to Bazel 0.25, you may first check if your codebase is compatible either by running [`bazelisk --migrate`](https://github.com/philwo/bazelisk) or by building your code with Bazel

0.24 and the following flags:

- [`--incompatible_disallow_legacy_py_provider`](https://github.com/bazelbuild/bazel/issues/7298)

- [`--incompatible_windows_escape_jvm_flags`](https://github.com/bazelbuild/bazel/issues/7486)

- [`--incompatible_remove_old_python_version_api`](https://github.com/bazelbuild/bazel/issues/7308) 

- [`--incompatible_allow_python_version_transitions`](https://github.com/bazelbuild/bazel/issues/7307) 

> - [`incompatible_disable_third_party_license_checking`](https://github.com/bazelbuild/bazel/issues/7553)

Reviewer 2: ```suggestion
- [`--incompatible_disable_third_party_license_checking`](https://github.com/bazelbuild/bazel/issues/7553)
```

- [`--incompatible_py3_is_default`](https://github.com/bazelbuild/bazel/issues/7359) and see [#7593](https://github.com/bazelbuild/issues/7593)

- [`--incompatible_disable_objc_library_resources`](https://github.com/bazelbuild/bazel/issues/7594) 

> - [`--incompatible_disable_legacy_crosstool_fields`](https://github.com/bazelbuild/bazel/issues/6861)

Reviewer 3: This is mentioned multiple times, probably because of rollback/roll forwards.

- [`--incompatible_remove_cpu_and_compiler_attributes_from_cc_toolchain`](https://github.com/bazelbuild/bazel/issues/7075) 

- [`--incompatible_use_toolchain_providers_in_java_common`](https://github.com/bazelbuild/bazel/issues/7186)

- [`--incompatible_remote_symlinks`](https://github.com/bazelbuild/bazel/issues/7917)

 

The following incompatible change flags have been added in this release and migration for them should begin. Check out the linked bugs to see when it is expected that the flags be turned on by default. 

- [`--incompatible_windows_native_test_wrapper`](https://github.com/bazelbuild/bazel/issues/6622)

- [`--incompatible_require_ctx_in_configure_features`](https://github.com/bazelbuild/bazel/issues/7793)

- [`--incompatible_do_not_split_linking_cmdline`](https://github.com/bazelbuild/bazel/issues/7687)

- [`--incompatible_string_join_requires_strings`](https://github.com/bazelbuild/bazel/issues/7802)

- [`--incompatible_use_python_toolchains`](https://github.com/bazelbuild/bazel/issues/7899) and see [#7375](https://github.com/bazelbuild/bazel/issues/7375)

> ## Android

Reviewer 2: is it me or is this changelist unusually short?

Author: If you take into account all of the flags, it's actually pretty long. So it is possible that people were strongly focusing on getting rid of things this release as opposed to adding new features.

 - Android NDK C++ toolchain is now configured in Starlark

 - Fix an issue where the Android resource processor did not surface

> 

Reviewer 2: remove these extra linebreaks, here and below

    errors from aapt2 compile and link actions.

## Other Changes

>   - Bazel is now ~20MiB smaller, from unbundling the Android rules'

Reviewer 2: may be elaborating more about the new binary size @meisterT

Reviewer 3: I think this is fine for now, we will have more comments the next two releases.
